File: harness/oracle/cases/help-doc-slug.ts
/**
 * P4.d6 oracle case — the help-document slug (v4 `lib/help/help-doc-slug.ts`
 * `helpDocSlug`, promoted out of `help-doc-sync.ts` by `d6e74145`).
 *
 * Tier 1: a pure function over a fixed corpus. Imports v4's REAL module (never
 * reimplements the chain) and emits one NDJSON row per input:
 *   { kind: 'slug', input, output }
 *
 * The corpus deliberately probes the parts a port gets wrong:
 *   - the two anchored, NON-global strips (`^help/`, `\.md$`) vs substring hits;
 *   - the GLOBAL `[^a-zA-Z0-9]` map (slashes, dots, spaces, underscores);
 *   - `toLowerCase()` AFTER the map;
 *   - ⚠ the non-/u regex's UTF-16 CODE-UNIT semantics: a non-BMP char is a
 *     surrogate pair and must yield TWO dashes, not one. A `chars()`-based port
 *     passes every ASCII case and fails only here.
 *
 * Run (Node 24, from the v4 checkout):
 *   N=~/.nvm/versions/node/v24.13.1/bin ; V5=<this worktree>
 *   cd ~/source/quilltap-server
 *   $N/node --import tsx $V5/harness/oracle/cases/help-doc-slug.ts > /tmp/oracle-help-doc-slug.ndjson
 */

const CORPUS: string[] = [
  // The documented example + the shipped help docs the drift was found through.
  'help/character-creation.md',
  'help/answer-confirmation.md',
  'help/brahma-console.md',
  'help/custom-tools.md',
  'help/post-office.md',
  'help/aurora.md',
  // Nested: only the FIRST `help/` is stripped; the inner slash maps to a dash.
  'help/nested/deep-doc.md',
  'help/a/b/c.md',
  'help/help/x.md',
  // Anchors are not substring matches.
  'docs/help/x.md',
  'help/a.mdx',
  'help/a.md.md',
  'help/a.mdx.md',
  'not-help/x.md',
  'helpful/x.md',
  // Case folding happens after the character map.
  'help/weird-CASE-Name.md',
  'help/Brahma Console.md',
  'help/ALLCAPS.md',
  // Assorted non-alphanumerics, each one dash.
  'help/a_b.c.md',
  'help/a+b=c.md',
  'help/spaced   out.md',
  'help/trailing-.md',
  'help/-leading.md',
  'help/dots...dots.md',
  'help/tab\there.md',
  'help/hash#frag.md',
  'help/paren(s).md',
  // Digits survive.
  'help/v2-migration-3.md',
  'help/123.md',
  // Degenerate.
  '',
  'help/',
  'help/.md',
  '.md',
  'help',
  '/',
  // ⚠ UTF-16 code-unit semantics — the surrogate-pair cases.
  'help/a\u{1F600}b.md', // one non-BMP char -> TWO dashes
  'help/\u{1F600}.md',
  'help/a\u{00E9}b.md', // BMP non-ASCII -> ONE dash
  'help/\u{4E2D}\u{6587}.md', // two BMP CJK units -> two dashes
  'help/mixed-\u{1F600}-\u{00E9}.md',
]

async function main(): Promise<void> {
  const { helpDocSlug } = await import('@/lib/help/help-doc-slug')

  for (const input of CORPUS) {
    process.stdout.write(
      JSON.stringify({ kind: 'slug', input, output: helpDocSlug(input) }) + '\n',
    )
  }
  process.exit(0)
}

main().catch((err) => {
  process.stderr.write(`help-doc-slug oracle failed: ${err?.stack ?? err}\n`)
  process.exit(1)
})

File: crates/quilltap-core/src/help_doc_slug.rs
//! Help-document slugs — port of v4 `lib/help/help-doc-slug.ts` (`helpDocSlug`),
//! promoted there by `d6e74145` out of `help-doc-sync.ts`'s private
//! `generateDocumentId`.
//!
//! **Why it exists (v4's rationale, carried forward):** the database primary key
//! for a help doc is a UUID that changes whenever the doc is re-created, so the
//! slug derived from its file path is the identifier everything OUTSIDE the
//! database uses — the Guide's category lists, `Related Pages` links between
//! docs, and the help-docs API.
//!
//! ## The consumers are NOT ported (a named deferral, not an oversight)
//!
//! v4 populates the slug in `HelpSearch`'s two mapping sites and matches it in
//! `getDocument(idOrSlug)` / `listDocuments()`. v5 has **no `HelpSearch` class
//! and no help-docs API** — `db::help_search` exposes only the two search
//! primitives, and the flat [`crate::db::help_search::HelpSearchResult`] is v4's
//! **tool-handler output** shape, which drops `slug`
//! (`lib/tools/handlers/help-search-handler.ts:64-71` maps `id`/`title`/`path`/
//! `url`/`score`/`content` only). Adding `slug` to that shape would DIVERGE from
//! v4's tool wire, so this module deliberately lands ahead of its consumers: it
//! is the promotion `d6e74145` performed, ready for the `HelpSearch` + help-docs
//! API surface when that lands.

/// v4 `helpDocSlug` — derive a help document's stable slug from its file path.
///
/// `help/character-creation.md` → `character-creation`.
///
/// v4's chain is
/// `.replace(/^help\//, '').replace(/\.md$/, '').replace(/[^a-zA-Z0-9]/g, '-').toLowerCase()`
/// — the two anchored strips are single, non-global replacements; the character
/// map is global.
///
/// **UTF-16 fidelity:** v4's `[^a-zA-Z0-9]` carries no `/u` flag, so it matches
/// per UTF-16 **code unit** — a non-BMP char (a surrogate PAIR) becomes **two**
/// dashes, not one. This port maps `encode_utf16()` for that reason rather than
/// `chars()`, which would emit a single dash and silently diverge. Every unit
/// that survives the map is ASCII-alphanumeric, so the result is pure ASCII and
/// the trailing `toLowerCase()` reduces to an ASCII lowercase (no ICU/locale
/// path can be reached from here).
pub fn help_doc_slug(rel_path: &str) -> String {
    // `.replace(/^help\//, '')` then `.replace(/\.md$/, '')` — anchored, once each.
    let stripped = rel_path.strip_prefix("help/").unwrap_or(rel_path);
    let stripped = stripped.strip_suffix(".md").unwrap_or(stripped);

    // `.replace(/[^a-zA-Z0-9]/g, '-')` per UTF-16 code unit, then `.toLowerCase()`.
    stripped
        .encode_utf16()
        .map(|unit| match u8::try_from(unit) {
            Ok(b) if b.is_ascii_alphanumeric() => (b as char).to_ascii_lowercase(),
            _ => '-',
        })
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn strips_help_prefix_and_md_suffix() {
        assert_eq!(
            help_doc_slug("help/character-creation.md"),
            "character-creation"
        );
        assert_eq!(help_doc_slug("help/aurora.md"), "aurora");
    }

    #[test]
    fn lowercases_and_maps_non_alphanumerics() {
        assert_eq!(help_doc_slug("help/Brahma Console.md"), "brahma-console");
        assert_eq!(help_doc_slug("help/weird-CASE-Name.md"), "weird-case-name");
        // Underscores and dots are non-alphanumeric → dashes.
        assert_eq!(help_doc_slug("help/a_b.c.md"), "a-b-c");
    }

    #[test]
    fn nested_paths_keep_their_separators_as_dashes() {
        // Only the FIRST `help/` is stripped (the regex is anchored, non-global);
        // the remaining slash falls to the character map.
        assert_eq!(help_doc_slug("help/nested/deep-doc.md"), "nested-deep-doc");
        assert_eq!(help_doc_slug("help/help/x.md"), "help-x");
    }

    #[test]
    fn anchors_are_not_substring_matches() {
        // `help/` only strips at the START; `.md` only at the END.
        assert_eq!(help_doc_slug("docs/help/x.md"), "docs-help-x");
        assert_eq!(help_doc_slug("help/a.mdx"), "a-mdx");
        // `.md` mid-path survives; the trailing one goes.
        assert_eq!(help_doc_slug("help/a.md.md"), "a-md");
    }

    #[test]
    fn non_bmp_char_yields_two_dashes_per_v4_utf16_semantics() {
        // U+1F600 is a surrogate PAIR in UTF-16; v4's non-/u regex replaces each
        // unit, so ONE emoji becomes TWO dashes.
        assert_eq!(help_doc_slug("help/a\u{1F600}b.md"), "a--b");
        // A BMP non-ASCII char is a single unit → a single dash.
        assert_eq!(help_doc_slug("help/a\u{00E9}b.md"), "a-b");
    }

    #[test]
    fn empty_and_degenerate_paths() {
        assert_eq!(help_doc_slug(""), "");
        assert_eq!(help_doc_slug("help/"), "");
        assert_eq!(help_doc_slug("help/.md"), "");
    }
}
